### DISCUSSION TITLE
WARNING messages from stat_analysis about multiple ALPHA values

### DISCUSSION THREAD

**Discussion by malloryprow** [dtcenter/METplus#2583] [Configuration] [answered]
I’m running StatAnalysis using the summary job.

> WARNING: For case "PRMSL:Z0:960000:ASIA:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:AUSTNZ:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:EURNAFR:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:NAMER:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:NHEM:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:NPOL:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:SHEM:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:SPOL:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.
> WARNING: For case "PRMSL:Z0:960000:TROPICS:NA", found 2 unique ALPHA values: 0.05000,-9999.00000.

I looked in the stat file and for the CNT line type the ALPHA value is 0.05, and for the GRAD line type the ALPHA value is NA (not -9999). I guess MET is doing something filling NA values with -9999.

I guess I’m not sure exactly how to get rid of the WARNING messages if the value in the stat file is NA but the NA value gets mapped to -9999.

I have put the related files on the FTP. I made the directory /incoming/irap/met_help/row_data but didn't cd into before I put the files on there. SIGH! I don't think I can move the files into row_data, but here is a list of the files: 

- Input: /incoming/irap/met_help/evs.stats.gfs.atmos.wmo.v20240501.stat
- Output: /incoming/irap/met_help/dump_row.stat, /incoming/irap/met_help/out_stat.stat
- Log: /incoming/irap/met_help/stat_analyis.log
- Config: /incoming/irap/met_help/STATAnalysisConfig_mallory

**Comment by JohnHalleyGotway** [comment, marked as answer]
Hi Mallory, let me apologize for the length of this up front! Here's the short answer... add `-by LINE_TYPE` to the job and the warnings go away. Although on my laptop, the runtime increases from 1:30 to 1:42:
```
-by LINE_TYPE,FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK,OBS_THRESH
```

Here's the gory details...

Thanks for passing along this data. I pulled it down, ran it through the latest development version of Stat-Analysis, and am able to replicate the behavior you describe... that warning messages are printed about multiple alpha values.

Looking at the job you're running, I see that it uses input from multiple line types (VCNT, CNT, GRAD):
```
jobs = [ "-job summary -column VCNT:RMSVE,VCNT:SPEED_ERR,CNT:ME,CNT:RMSE,CNT:ANOM_CORR,CNT:MAE,CNT:RMSFA,CNT:RMSOA,CNT:FSTDEV,CNT:OSTDEV,GRAD:S1 -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK,OBS_THRESH -dump_row dump_row.stat -out_stat out.stat" ];
```
And those multiple input line types are definitely involved here. I'm used to running each job for a single input line type, but this syntax is certainly supported.

I started by inventorying the input data. `evs.stats.gfs.atmos.wmo.v20240501.stat` contains 1,675,920 lines of data consisting of 590,453 CNT lines, 650,689 VCNT lines, 434,417 MCTC lines, 360 GRAD lines, and 1 header line.

The `dump_row.stat` file written by the job contains the input lines that were ACTUALLY used by this job. And for this job, it only contains 6,481 lines of data consisting of 3960 CNT, 2160 VCNT, 360 GRAD, and 1 header line.

Checking your config file, I see that those are the 6,480 .stat lines for `gfs` and `grid2grid_upperair`.

I tried splitting up your 1 big job into 3 separate jobs, one for each input line type. That produced the following runtime error:
```
ERROR  : dump_stat_line() -> unexpected line type value VCNT
```
This bug issue is easily fixed, which I did in my local copy and will include in beta5. But for now, I just reran _WITHOUT_ the `-dump_row` option. Running the following does work and does not trigger any warnings:
```
/d1/personal/johnhg/MET/MET_development/MET-develop/bin/stat_analysis \
-lookin evs.stats.gfs.atmos.wmo.v20240501.stat \
-config STATAnalysisConfig_jhg \
-log run_sa_jhg.log \
-out wmo_stats.txt
```
With these jobs:
```
jobs = [ "-job summary -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK,OBS_THRESH -line_type VCNT -column RMSVE,SPEED_ERR",
         "-job summary -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK,OBS_THRESH -line_type CNT  -column ME,RMSE,ANOM_CORR,MAE,RMSFA,RMSOA,FSTDEV,OSTDEV",
         "-job summary -by FCST_VAR,FCST_LEV,FCST_LEAD,VX_MASK,OBS_THRESH -line_type GRAD -column S1" ];
```
And I'm using the `-out` command line option to write to `wmo_stats.txt` rather then using `-out_stat` for each job. Note that the `-job summary` does NOT write any .stat output. So that's why the `-out_stat` output file is empty. Instead, the summary job prints info to the screen or redirects it to the output file specified using `-out`.

We could consider making the output from `-job summary` a fully supported STAT line that could be loaded into a METdataio database. But that's currently not the case, so you're stuck with the output just in a text file.

Here's the resulting [wmo_stats.txt](https://github.com/dtcenter/METplus/files/15311812/wmo_stats.txt) output file I get. I confirmed that the log file contains no warnings. I'll note that it takes about 1.5 - 2.0 minutes to run this on my Mac laptop. I'm curious how long this takes on WCOSS2 since reading/writing from your temp directory (`/lfs/h2/emc/stmp/mallory.row/stat_analysis`) is much slower than me reading/writing from `/tmp` on my laptop. Is it much slower?

I checked the log messages for...
- Running 1 big job:
```
DEBUG 2: Computing output for 3060 case(s).
DEBUG 2: Job 1 used 6480 out of 6480 STAT lines.
```
- Running 1 job for each input line type...
```
DEBUG 2: Computing output for 1080 case(s).
DEBUG 2: Job 1 used 2160 out of 6480 STAT lines.
...
DEBUG 2: Computing output for 1980 case(s).
DEBUG 2: Job 2 used 3960 out of 6480 STAT lines.
...
DEBUG 2: Computing output for 180 case(s).
DEBUG 2: Job 3 used 360 out of 6480 STAT lines.
```
And I did confirm that both approaches produce the same 18,180 `SUMMARY` output lines... and I spot-checked a couple to make sure the numbers match.

Here's some interesting counts:
- 1 big job reports `3060` cases.
- 3 separate jobs report `1080 + 1980 + 180 = 3240` cases.
- Your log file contains `180` warning messages.
- Interestingly, `3240 - 180 = 3060`. So you're getting warnings about those "extra" 180 cases which contain data for both the GRAD and CNT lines.

That made me realize we can just add `-by LINE_TYPE` to your original job to define all 3240 cases and avoid those warnings altogether. With that change, here's the log messages:
```
DEBUG 2: Computing output for 3240 case(s).
DEBUG 2: Job 1 used 6480 out of 6480 STAT lines.
```
So that's an easy change I'd recommend without needing to modify the logic of Stat-Analysis at all.

**Reply by JohnHalleyGotway** [reply]
FYI, here's the bugfix issue for the `-dump_row dump_vcnt.txt -line_type VCNT` problem I ran into:
https://github.com/dtcenter/MET/issues/2888

**Comment by malloryprow** [comment]
Thanks for that answer John! I'm glad that this all uncovered a bug fix and that was easy. I'm not sure how exactly long it took maybe a minute or two more than it took you (I was testing in the stmp space and it was filling up yesterday so I had to clean things up after I sent everything to the FTP).
